pkg/proxy/proxy/transparent.go

```diff
@@ -175,6 +203,55 @@ func (t *TransparentProxyService) ServeMetadata(addr string) {
 	}
 }
 
+// Check that the requested url is allowed by the network policy.
+func (proxy TransparentProxyService) CheckNetworkPolicy(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
+	if proxy.Mode != EnforcementMode {
+		return req, nil
+	}
+	url := req.URL
+	if proxy.Policy != nil {
+		for _, rule := range proxy.Policy.Rules {
+			if url.Hostname() != rule.Host {
+				continue
+			}
+			if urlSatisfiesRule(url, rule) {
+				return req, nil
+			}
+		}
+	}
+	log.Printf("Request to %s blocked by network policy", url.String())
+	errorMessage := fmt.Sprintf("Access to %s is blocked by the proxy's network policy", url.String())
+	return req, goproxy.NewResponse(req, goproxy.ContentTypeText, http.StatusForbidden, errorMessage)
+}
+
+type MatchingStrategy int
+
+const (
+	PathPrefix MatchingStrategy = iota
+	FullPath
+)
+
+type NetworkPolicy struct {
+	Rules []NetworkPolicyRule
+}
+
+type NetworkPolicyRule struct {
```

Comment:
I think the existing goproxy methods for filtering are not quite as flexible as we may need them to be. My initial thought is that it may not necessarily make the code simpler/cleaner since we will potentially need custom handlers for things like redirection in the future anyways, but this could be something to look into.

Reply:
yeah I sort of suspected that but i figured it was worth throwing out there. thanks!